[cond] support unbacked symbool inputs

ydwu4 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #133589

Fixes https://github.com/pytorch/pytorch/issues/133577.

In dynamo, when received an unbacked symbool input, we create an unbacked symint to replace it.

The alternative approach of `not realizing the pred LazyVariable in cond` doesn't work because we need to get the proxy of the symbool input.


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

ydwu4 committed:
[cond] support unbacked symbool inputs

[ghstack-poisoned]

ezyang commented on torch/_dynamo/variables/builder.py:
```diff
@@ -861,14 +861,22 @@ def build_key_value(i, k, v):
+            else:
+                # We need to create an unbacked symint to replace the unbacked symbool.
+                # Call ignore_fresh_unbacked_symbols to skip the pending unbacked symbols check.
+                with self.tx.output.shape_env.ignore_fresh_unbacked_symbols():
```
This seems suspicious. I can see, at least mechanically, why this is helpful: the unbacked symbol you generate here never ends up getting bound to a symnode. But... maybe it should? In particular, we do create a graph input immediately below, which unbacked_bindings could be set on. I'm not sure though...

ydwu4 replied:
em... I can call set_example_value after we create the graph input proxy to bind the unbacked symint to a path, which looks like {u0: ()}. I'm not sure if this is correct or not because technically the symint is converted from a symbool. Also not sure how to add a better test for this. Update the diff for more comments.

ydwu4 committed:
Update on "[cond] support unbacked symbool inputs"


Fixes https://github.com/pytorch/pytorch/issues/133577.

In dynamo, when received an unbacked symbool input, we create an unbacked symint to replace it.

The alternative approach of `not realizing the pred LazyVariable in cond` doesn't work because we need to get the proxy of the symbool input.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

ezyang reviewed (approved):
Kinda spooky but if it works we can roll with it.

ydwu4 commented:
@pytorchbot merge
